rust/extns/numaflow-sqs/src/source.rs

```diff
@@ -0,0 +1,706 @@
+                }
+            };
+            if let Err(err) = self
+                .client
+                .delete_message()
+                .queue_url(self.queue_url.clone())
+                .receipt_handle(offset)
+                .send()
+                .await
+            {
+                tracing::error!(
+                    ?err,
+                    "{} {}",
+                    self.queue_url.clone(),
+                    "Error while deleting message from SQS"
+                );
+                return Err(Error::SQS(err.into()));
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn get_pending_messages(&mut self, queue_name: String) -> Result<Option<usize>> {
+        let get_queue_url_response = self.get_queue_url(queue_name.clone()).await;
+
+        if let Err(err) = get_queue_url_response {
+            tracing::error!(
+                ?err,
+                queue_name = queue_name.clone(),
+                "Failed to get queue url"
+            );
+            return Err(err);
+        }
+
+        let sdk_response = self
+            .client
+            .get_queue_attributes()
+            .queue_url(self.queue_url.clone())
+            .attribute_names(aws_sdk_sqs::types::QueueAttributeName::ApproximateNumberOfMessages)
+            .send()
+            .await;
+
+        let get_queue_attributes_output = match sdk_response {
+            Ok(output) => output,
+            Err(err) => {
+                tracing::error!(
+                    ?err,
+                    queue_url = self.queue_url,
+                    "Failed to get queue attributes from SQS"
+                );
+                return Err(Error::SQS(err.into()));
+            }
+        };
+
+        let attributes = match get_queue_attributes_output.attributes {
+            Some(attributes) => attributes,
+            None => return Ok(None),
+        };
+
+        let value = match attributes.get(&QueueAttributeName::ApproximateNumberOfMessages) {
+            Some(value) => value,
+            None => return Ok(None),
+        };
+
+        let approx_pending_messages_count = match value.parse::<usize>() {
+            Ok(count) => count,
+            Err(err) => {
+                tracing::error!(?err, "Failed to parse ApproximateNumberOfMessages");
+                return Err(Error::Other(
+                    "Failed to parse ApproximateNumberOfMessages".to_string(),
+                ));
+            }
+        };
+        Ok(Some(approx_pending_messages_count))
+    }
+}
+
+/// Public interface for interacting with SQS queues.
+///
+/// Design principles:
+/// - Thread-safe through actor model
+/// - Configurable batch sizes and timeouts
+/// - Clean abstraction of SQS complexity
+/// - Efficient message processing
+#[derive(Clone)]
+pub struct SQSSource {
+    batch_size: usize,
+    /// timeout for each batch read request
+    timeout: Duration,
+    actor_tx: mpsc::Sender<SQSActorMessage>,
+    queue_name: String,
+}
```

Comment:
can we avoid cloning the whole struct? `actor_tx` is all that is required to be cloned (i think).